BB-432: remove useless kafka connectors

welansari opened the pull request:
Issue: BB-432

- Avoid spawning the connectors at startup
- Delete connectors when no bucket is assigned to them

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/backbeat/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-432 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.6.27`

* `8.7.0`


Please check the `Fix Version/s` of BB-432, or the target
branch of this pull request.

Commit:
destroy kafka connect connectors when no bucket is assigned to them

Issue: BB-432

Commit:
avoid creating connectors at startup

Issue: BB-432

Commit:
test destroying kafka connect connectors when no bucket assigned to them

Issue: BB-432

bert-e commented:
# History mismatch



Merge commit #8e246f11d5db47a17e5a1eb9962fa3ca01fd235a on the integration branch
`w/8.7/improvement/BB-432-remove-useless-kafka-connectors` is merging a branch which is neither the current
branch `improvement/BB-432-remove-useless-kafka-connectors` nor the development branch
`development/8.7`.

It is likely due to a rebase of the branch `improvement/BB-432-remove-useless-kafka-connectors` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Kerkesni commented:
/reset

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/BB-432-remove-useless-kafka-connectors` into
`development/8.6`
* [w/8.7/improvement/BB-432-remove-useless-kafka-connectors](https://github.com/scality/backbeat/compare/development/8.7...w/8.7/improvement/BB-432-remove-useless-kafka-connectors) will be merged into `development/8.7` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

bert-e commented:
# History mismatch



Merge commit #8491ca92695ad0c83bf582c632dcf8bef40c23e1 on the integration branch
`w/8.7/improvement/BB-432-remove-useless-kafka-connectors` is merging a branch which is neither the current
branch `improvement/BB-432-remove-useless-kafka-connectors` nor the development branch
`development/8.7`.

It is likely due to a rebase of the branch `improvement/BB-432-remove-useless-kafka-connectors` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Kerkesni commented:
/reset

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/BB-432-remove-useless-kafka-connectors` into
`development/8.6`
* [w/8.7/improvement/BB-432-remove-useless-kafka-connectors](https://github.com/scality/backbeat/compare/development/8.7...w/8.7/improvement/BB-432-remove-useless-kafka-connectors) will be merged into `development/8.7` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -238,47 +223,97 @@ class ConnectorsManager {
+                    connector: connector.name
+                });
+                return true;
+            }
```
would it not simplify the code [below] if there is an additional `else` here?

```
else if (connector.isRunning) {
   return await connector.updatePipeline(true);
}
```

francoisferrand commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -105,40 +105,24 @@ class ConnectorsManager {
+            kafkaConnectHost: this._kafkaConnectHost,
+            kafkaConnectPort: this._kafkaConnectPort,
+        });
+        return connector;
```
not sure how to handle `onConnectorsInstantiated` metrics, now that the connectors are not created on startup :-/
it seems to me we are interested in the number of connectors created & destroyed (as 2 counters, when we actually creaste/destroy one connector), but not sure how it relates to this one.

francoisferrand commented on extensions/oplogPopulator/modules/Connector.js:
```diff
@@ -6,6 +6,7 @@ const connectorParams = joi.object({
     name: joi.string().required(),
     config: joi.object().required(),
     buckets: joi.array().required(),
+    isRunning: joi.boolean().required(),
```
`isRunning` is good enough for now, as long as the allocator is always give a _static_ list of connectors.

however, we should (eventually, not here) implement a mecanism to create more connectors when we reach a limit (max mongo query size, pre-configured number of collections...) : and at that point the allocator should be able to allocate more connectors if needed. But this is not for today.

francoisferrand commented on tests/unit/oplogPopulator/Connector.js:
```diff
@@ -40,20 +41,25 @@ describe('Connector', () => {
         it('Should destroy connector', async () => {
             const deleteStub = sinon.stub(connector._kafkaConnect, 'deleteConnector')
                 .resolves();
+            connector._isRunning = true;
+            assert.strictEqual(connector.isRunning, true);
             await connector.destroy();
```
what is the point of this assert?

welansari replied:
makes sure the getter works properly

francoisferrand replied:
oh yes, i missed the missing underscore and getter...

francoisferrand commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -105,40 +105,24 @@ class ConnectorsManager {
-            });
-            throw errors.InternalError.customizeDescription(err.description);
-        }
+    addConnector() {
```
function is not async anymore, so call site should be updated:

```js
const newConnector = await this.addConnector();
                     ^^^^^
```

francoisferrand reviewed (commented):
looks mostly good, though a deeper refactoring should be done (eventually) to make the list really dynamic instead of virtually pre-creating it to a specified length.

bert-e commented:
# History mismatch



Merge commit #dce19c1ed4cedab4b8c8364b6ee440007f187ae4 on the integration branch
`w/8.7/improvement/BB-432-remove-useless-kafka-connectors` is merging a branch which is neither the current
branch `improvement/BB-432-remove-useless-kafka-connectors` nor the development branch
`development/8.7`.

It is likely due to a rebase of the branch `improvement/BB-432-remove-useless-kafka-connectors` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Kerkesni commented:
/reset

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/BB-432-remove-useless-kafka-connectors` into
`development/8.6`
* [w/8.7/improvement/BB-432-remove-useless-kafka-connectors](https://github.com/scality/backbeat/compare/development/8.7...w/8.7/improvement/BB-432-remove-useless-kafka-connectors) will be merged into `development/8.7` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

bert-e commented:
# History mismatch



Merge commit #5e4550d211d81f738b74702d29233f379d9d61c5 on the integration branch
`w/8.7/improvement/BB-432-remove-useless-kafka-connectors` is merging a branch which is neither the current
branch `improvement/BB-432-remove-useless-kafka-connectors` nor the development branch
`development/8.7`.

It is likely due to a rebase of the branch `improvement/BB-432-remove-useless-kafka-connectors` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Kerkesni commented:
/reset

bert-e commented:
# History mismatch



Merge commit #5e4550d211d81f738b74702d29233f379d9d61c5 on the integration branch
`w/8.7/improvement/BB-432-remove-useless-kafka-connectors` is merging a branch which is neither the current
branch `improvement/BB-432-remove-useless-kafka-connectors` nor the development branch
`development/8.7`.

It is likely due to a rebase of the branch `improvement/BB-432-remove-useless-kafka-connectors` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/BB-432-remove-useless-kafka-connectors` into
`development/8.6`
* [w/8.7/improvement/BB-432-remove-useless-kafka-connectors](https://github.com/scality/backbeat/compare/development/8.7...w/8.7/improvement/BB-432-remove-useless-kafka-connectors) will be merged into `development/8.7` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand commented on extensions/oplogPopulator/OplogPopulatorMetrics.js:
```diff
@@ -132,6 +132,22 @@ class OplogPopulatorMetrics {
+     */
+    onConnectorDestroyed() {
+        try {
+            this.connectors.dec({ isOld: false }, 1);
```
this is not correct: we may remove either old or new connector : the "old" param is based on how the connector was created  (retrieved from kafka-connect or explcitely recreated), so must be tracked throughout...

francoisferrand commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -238,47 +222,98 @@ class ConnectorsManager {
+        try {
+            if (connector.isRunning && connector.bucketCount === 0) {
+                await connector.destroy();
+                this._metricsHandler.onConnectorDestroyed(false);
```
there is no param in this function....
and if it were, it could be either true or false depending on how that specific connector was created...

Kerkesni commented:
/wait

welansari commented:
/approve

francoisferrand committed:
Remove isOld from connectors metrics

This label is not really useful, and not used in dashboard anyway. So
best remove it instead of adding more logic to maintain it.

IssyeL BB-432

nicolas2bert commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -220,9 +205,8 @@ class ConnectorsManager {
             const nbConnectorsToAdd = this._nbConnectors - this._connectors.length;
             await timesLimit(nbConnectorsToAdd, 10, async () => {
-                const newConnector = await this.addConnector();
+                const newConnector = this.addConnector();
```
Nothing is asynchronous here. What is the need for `timesLimit`? It could be misleading.

francoisferrand replied:
i guess it is a remain from the origin (async) code, i'll check

nicolas2bert commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -238,47 +222,98 @@ class ConnectorsManager {
+            if (connector.isRunning && connector.bucketCount === 0) {
+                await connector.destroy();
+                this._metricsHandler.onConnectorDestroyed();
+                this._logger.info('Successfully spawned a connector', {
```
Log should be `Successfully destroyed a connector`? 
Next condition to be updated as well.

nicolas2bert commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -238,47 +222,98 @@ class ConnectorsManager {
+                this._logger.info('Successfully destroyed a connector', {
+                    method: 'ConnectorsManager._spawnOrDestroyConnector',
+                    connector: connector.name
+                });
```
Should we not also update the connector pipeline after spawning?

francoisferrand replied:
updating after should not be needed (the `config` is passed to creation api), but indeed `config.pipeline` should be updated before making the call...

francoisferrand replied:
ok we are good: `updatePipeline()` is called automatically by `addBucket()` and `removeBucket()`, and will update the pipeline field automatically. However, it is called with the `doUpdate` param set to `false`, in order to only update in memory : then the config is either applied by the create all (which passes the struct) or by an explicit extra call to `updatePipeline(true)` when the connector was already there.

nicolas2bert commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -238,47 +222,98 @@ class ConnectorsManager {
+     * @returns {Promise<Boolean>} true if connector state changed
+     * @throws {InternalError}
+     */
+    async _spawnOrDestroyConnector(connector) {
```
Now that bucket can be allocated to non-running connector(s) (before connector gets spawn), could this have an impact on the design? (e.g. potentially missing some MongoDB changes until the connector gets spawned...?)

francoisferrand replied:
not really I think, we anyway have a delay between the "activation of new feature which requires oplog" and the moment when a connector is indeed running (i.e. the time for the oplog populator to setup everything) : it may only increase the delay ever so slightly...

but the immediate benefit is that we avoid running any change stream in Veeam/backup use case, and thus avoid risk of breaking the platform.

francoisferrand committed:
Bump release 8.6.27

Issue: BB-432

francoisferrand commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/BB-432-remove-useless-kafka-connectors` into
`development/8.6`
* [w/8.7/improvement/BB-432-remove-useless-kafka-connectors](https://github.com/scality/backbeat/compare/development/8.7...w/8.7/improvement/BB-432-remove-useless-kafka-connectors) will be merged into `development/8.7` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/backbeat/commit/69308957176acc7070b23585d549bd5e06b57581) did not succeed in branch w/8.7/improvement/BB-432-remove-useless-kafka-connectors.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`



Please check the status of the associated issue BB-432.

Goodbye kerkesni.
